lib/fluent/plugin/auditd.rb

```diff
@@ -1,120 +1,113 @@
-            end
-            state = states[:start]
-          elsif line[i] == '$'
-            # end of input
-            if stack.last == '$'
-              insert(result, nested, key, nested_key, value) unless value == '?'
-            else
-              # input symbol doesn't match stack symbol
-              handle_error line, result
-          end
-          else
-            # reading value
-            value << line[i]
-          end
+      vc = {}
+      if metadata = /(?<g1>.*?) msg='(?<g2>.*?)'/.match(line) 
+        if (!metadata['g1'].nil?) && (!metadata['g2'].nil?)
+          parse_metadata(result, metadata['g1'].split)
+          parse_msg(vc, metadata['g2'].split)
+          result['virt_control'] = vc
         else
-          # unknown state
-          handle_error line, result
+          handle_error "Couldn't parse message: #{line}"
         end
+      else
+        handle_error "Couldn't parse message: #{line}"
       end
-      return result
+      return normalize(result)
     end
 
-    def handle_error(line, result)
-      raise AuditdParserException, (sprintf \
-        "Error after processing string '%s'\nBuilt %s", line, result)
+    def parse_metadata(result, metadata)
+      result[TIME] = metadata[1].sub(/msg=audit\((?<g1>.*):\d+\):/, '\k<g1>')
+      for i in 2...metadata.length
+        pair = metadata[i].split('=')
+        insert_or_merge(result, pair[0], pair[1]) unless pair[1].nil? or pair[1] == '?'
+      end
     end
 
-    def insert(result, nested, key, nested_key, value)
-      # auditd may duplicate keys, save timestamp before it can be overriden
-      if key == MSG and value.start_with?(AUDIT)
-        key = TIMESTAMP
-        value.sub!(/audit\((?<g1>.*):\d{4,}\):/, '\k<g1>')
+    def parse_msg(result, msg)
+      msg.each do |part|
+        pair = part.split('=')
+        insert_or_merge(result, pair[0], pair[1]) unless pair[1].nil? or pair[1] == '?'
       end
+    end
 
-      if nested
-        if result[nested_key][key].nil?
-          result[nested_key][key] = value
-        else
-          if result[nested_key][key].kind_of?(Array)
-            result[nested_key][key] << value
-          else
-            temp = result[nested_key][key]
-            result[nested_key][key] = [temp, value]
-          end
-        end
+    def insert_or_merge(result, key, value)
+      if result[key].nil?
+        result[key] = value
+      elsif result[key].kind_of?(Array)
+        result[key] << value
       else
-        if result[key].nil?
-          result[key] = value
-        else
-          if result[key].kind_of?(Array)
-            result[key] << value
-          else
-            temp = result[key]
-            result[key] = [temp, value]
-          end
-        end
+        temp = result[key]
+        result[key] = [value, temp]
       end
     end
+
+    def handle_error(msg)
+      raise AuditdParserException, msg
+    end
+
+    def normalize(target)
+      event = {}
```

Comment:
You may remove all of the unless statements and sanitize the resulting hash with `compact`.  This will remove all of the keys that have nil values